src/argon2_fmt_plug.c

```diff
@@ -245,6 +266,35 @@ static void set_salt(void *salt)
 	memcpy(&saved_salt, salt, sizeof(struct argon2_salt));
 }
 
+static int allocate(uint8_t **memory, size_t size)
+{
+	if (thread_mem[THREAD_NUMBER].used)
+		error_msg("%s() thread %u: Memory allocated twice\n", __FUNCTION__, THREAD_NUMBER);
+
+	if (thread_mem[THREAD_NUMBER].ptr == NULL) {
+		thread_mem[THREAD_NUMBER].ptr = mem_alloc(size);
+		thread_mem[THREAD_NUMBER].size = size;
+	} else if (thread_mem[THREAD_NUMBER].size < size) {
+		thread_mem[THREAD_NUMBER].ptr = mem_realloc(thread_mem[THREAD_NUMBER].ptr, size);
+		thread_mem[THREAD_NUMBER].size = size;
+	}
+
+	thread_mem[THREAD_NUMBER].used = 1;
+	*memory = thread_mem[THREAD_NUMBER].ptr;
+
+	return 1;
+}
+
+static void deallocate(uint8_t *memory, size_t size)
+{
+	if (!thread_mem[THREAD_NUMBER].used)
+		error_msg("%s(): thread %u: Freed memory not in use\n", __FUNCTION__, THREAD_NUMBER);
+	if (thread_mem[THREAD_NUMBER].size < size)
+		error_msg("%s(): thread %u: incorrect size %zu, was %zu\n", __FUNCTION__, THREAD_NUMBER, size, thread_mem[THREAD_NUMBER].size);
+
+	thread_mem[THREAD_NUMBER].used = 0;
+}
```

Comment:
Not sure why they supply a size to `deallocate()` but I used it for an extra check, bc why not.

Reply:
Maybe can even insist on `== size` in that check?

Reply:
No, it might be smaller after we reused the existing alloc for a smaller request.